agent/src/main/java/com/microsoft/applicationinsights/agent/internal/agent/CustomClassWriter.java

```diff
@@ -0,0 +1,48 @@
+package com.microsoft.applicationinsights.agent.internal.agent;
+
+
+import org.objectweb.asm.ClassWriter;
+
+public class CustomClassWriter extends ClassWriter {
+
+
+    ClassLoader classLoader;
+
+    public CustomClassWriter(int writerFlag, ClassLoader loader)
+    {
+        super(writerFlag);
+        this.classLoader = loader;
+    }
+
+    protected String getCommonSuperClass(String className1, String className2)
+    {
+        Class class1;
+        Class class2;
+        try
+        {
+            class1 = Class.forName(className1.replace('/', '.'), false, this.classLoader);
+            class2 = Class.forName(className2.replace('/', '.'), false, this.classLoader);
+        }
+        catch (Exception th) {
+            throw new RuntimeException(th.getMessage());
+        }
+
+        if (class1.isAssignableFrom(class2)) {
+            return className1;
+        }
+        if (class2.isAssignableFrom(class1)) {
+            return className2;
+        }
+
+        if ((class1.isInterface()) || (class2.isInterface())) {
+            return "java/lang/Object";
+        }
+
+        do {
+            class1 = class1.getSuperclass();
+        }
+        while (!(class1.isAssignableFrom(class2)));
```

Comment:
Any possible issue with endless cycle or exception here if class1 has no more superClass and class2 happens to be not assignable (e.g. null or something...)?

Reply:
@Dmitry-Matveev to be specific about this, this is the exact implementation as how ASM does by default. Only reason to override this method was because ASM uses Class.getClassLoader() method to load the current class and therefore it gets the ClassLoader which doesn't have all the required classes loaded in it. In this override I pass the ClassLoader from the Callee in order to avoid the lookup and have the classloader which has all the required classes loaded. Therefore I believe that since there is minimal change done from ASM implementation this possibility is very narrow. Is there a way we can test this? How can we know if there is a null being passed or something like that?

Reply:
@dhaval24 To test this method, you could refactor it into a `public static` utility method which takes the ClassLoader along with the other parameters and write tests against that.